File: research/quantfoundry-signals/src/quantfoundry_signals/symbol.py
"""Canonical symbol parser/formatter — mirrors ``server/signals/symbol.ts``.

A symbol is a canonical, colon-separated string. All signal-side symbol
handling goes through :func:`parse_symbol` and :func:`format_symbol`;
ad-hoc string munging on incoming/outgoing symbol fields is a bug.

Grammar (from ``docs/tdd/signal-ingress.md §2.1.1``):

* ``EQ:<ticker>`` — e.g. ``EQ:SPY``
* ``OPT:<root>:<expiry>:<right>:<strike>`` — e.g. ``OPT:SPY:2026-01-16:C:500``
* ``FUT:<root>:<contract>`` — e.g. ``FUT:ES:2026-06``
* ``FOP:<root>:<contract>:<expiry>:<right>:<strike>``
  — e.g. ``FOP:ES:2026-06:2026-05-23:C:5000``
* ``V:<label>`` — virtual (non-tradeable), e.g. ``V:regime-spx-vol``

Roots/tickers/contracts: ``[A-Z0-9_-]+`` (uppercase). Dates: ``YYYY-MM-DD``.
Contract months for FUT/FOP accept ``YYYY-MM`` or the same date form. Rights:
``C``/``P``. Strikes: positive finite numbers, no trailing zeros. Virtual
labels are case-insensitive at parse time (per the TS implementation).
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

# ── Constants ──────────────────────────────────────────────────────

VALID_CLASSES = frozenset({"EQ", "OPT", "FUT", "FOP", "V"})
_TOKEN_RE = re.compile(r"^[A-Z0-9_-]+$")
_DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")
_MONTH_RE = re.compile(r"^\d{4}-\d{2}$")
_RIGHT_RE = re.compile(r"^[CP]$")
_V_LABEL_RE = re.compile(r"^[A-Za-z0-9_-]+$")

SymbolClass = Literal["EQ", "OPT", "FUT", "FOP", "V"]
Right = Literal["C", "P"]


# ── Parsed-symbol records ──────────────────────────────────────────


@dataclass(frozen=True)
class EqSymbol:
    ticker: str
    cls: Literal["EQ"] = "EQ"


@dataclass(frozen=True)
class OptSymbol:
    root: str
    expiry: str
    right: Right
    strike: float
    cls: Literal["OPT"] = "OPT"


@dataclass(frozen=True)
class FutSymbol:
    root: str
    contract: str
    cls: Literal["FUT"] = "FUT"


@dataclass(frozen=True)
class FopSymbol:
    root: str
    contract: str
    expiry: str
    right: Right
    strike: float
    cls: Literal["FOP"] = "FOP"


@dataclass(frozen=True)
class VSymbol:
    label: str
    cls: Literal["V"] = "V"


ParsedSymbol = EqSymbol | OptSymbol | FutSymbol | FopSymbol | VSymbol


# ── Parse ──────────────────────────────────────────────────────────


def parse_symbol(s: str) -> ParsedSymbol:
    """Parse a canonical symbol string into its typed record.

    Raises :class:`ValueError` on any malformed input — same error
    surface as the TS implementation, with messages chosen to be
    diff-friendly so cross-language regressions are easy to spot.
    """
    if not isinstance(s, str) or not s:
        raise ValueError("Symbol must be a non-empty string")

    parts = s.split(":")
    cls = parts[0]
    if cls not in VALID_CLASSES:
        raise ValueError(f"Unknown symbol class: {cls}")

    if cls == "EQ":
        if len(parts) != 2:
            raise ValueError(f"EQ symbol must have 2 parts, got {len(parts)}: {s}")
        ticker = parts[1]
        if not _TOKEN_RE.match(ticker):
            raise ValueError(f"Invalid ticker: {ticker}")
        return EqSymbol(ticker=ticker)

    if cls == "OPT":
        if len(parts) != 5:
            raise ValueError(f"OPT symbol must have 5 parts, got {len(parts)}: {s}")
        _, root, expiry, right, strike_str = parts
        if not _TOKEN_RE.match(root):
            raise ValueError(f"Invalid root: {root}")
        if not _DATE_RE.match(expiry):
            raise ValueError(f"Invalid expiry date: {expiry}")
        if not _RIGHT_RE.match(right):
            raise ValueError(f"Invalid right: {right}")
        strike = _parse_strike(strike_str)
        return OptSymbol(root=root, expiry=expiry, right=right, strike=strike)  # type: ignore[arg-type]

    if cls == "FUT":
        if len(parts) != 3:
            raise ValueError(f"FUT symbol must have 3 parts, got {len(parts)}: {s}")
        _, root, contract = parts
        if not _TOKEN_RE.match(root):
            raise ValueError(f"Invalid root: {root}")
        if (
            not _TOKEN_RE.match(contract)
            and not _DATE_RE.match(contract)
            and not _MONTH_RE.match(contract)
        ):
            raise ValueError(f"Invalid contract: {contract}")
        return FutSymbol(root=root, contract=contract)

    if cls == "FOP":
        if len(parts) != 6:
            raise ValueError(f"FOP symbol must have 6 parts, got {len(parts)}: {s}")
        _, root, contract, expiry, right, strike_str = parts
        if not _TOKEN_RE.match(root):
            raise ValueError(f"Invalid root: {root}")
        if not _DATE_RE.match(expiry):
            raise ValueError(f"Invalid expiry date: {expiry}")
        if not _RIGHT_RE.match(right):
            raise ValueError(f"Invalid right: {right}")
        strike = _parse_strike(strike_str)
        return FopSymbol(
            root=root,
            contract=contract,
            expiry=expiry,
            right=right,  # type: ignore[arg-type]
            strike=strike,
        )

    # cls == "V"
    if len(parts) != 2:
        raise ValueError(f"V symbol must have 2 parts, got {len(parts)}: {s}")
    label = parts[1]
    if not _V_LABEL_RE.match(label):
        raise ValueError(f"Invalid virtual label: {label}")
    return VSymbol(label=label)


def _parse_strike(s: str) -> float:
    try:
        v = float(s)
    except ValueError as exc:
        raise ValueError(f"Invalid strike: {s}") from exc
    if v <= 0 or v != v or v == float("inf") or v == float("-inf"):
        raise ValueError(f"Invalid strike: {s}")
    return v


# ── Format ─────────────────────────────────────────────────────────


def format_symbol(sym: ParsedSymbol) -> str:
    """Render a :class:`ParsedSymbol` back to its canonical string."""
    if isinstance(sym, EqSymbol):
        return f"EQ:{sym.ticker}"
    if isinstance(sym, OptSymbol):
        return f"OPT:{sym.root}:{sym.expiry}:{sym.right}:{_format_strike(sym.strike)}"
    if isinstance(sym, FutSymbol):
        return f"FUT:{sym.root}:{sym.contract}"
    if isinstance(sym, FopSymbol):
        return (
            f"FOP:{sym.root}:{sym.contract}:{sym.expiry}:"
            f"{sym.right}:{_format_strike(sym.strike)}"
        )
    # VSymbol — exhaustive over the union, but mypy doesn't narrow.
    return f"V:{sym.label}"


def _format_strike(strike: float) -> str:
    """Format a strike with no trailing zeros (500 stays "500", 4787.5 stays
    "4787.5"). Mirrors the TS ``String(strike)`` behaviour for finite floats."""
    if strike == int(strike):
        return str(int(strike))
    return str(strike)


__all__ = [
    "VALID_CLASSES",
    "EqSymbol",
    "FopSymbol",
    "FutSymbol",
    "OptSymbol",
    "ParsedSymbol",
    "Right",
    "SymbolClass",
    "VSymbol",
    "format_symbol",
    "parse_symbol",
]

File: server/symbols/symbol.ts
// ── Canonical Symbol Parser/Formatter ──────────────────────────────
// One parser, one formatter. All symbol handling goes through this.
// Defined in: docs/tdd/signal-ingress.md, topic 1

import type { ParsedSymbol, SymbolClass } from "../../src/types/symbols.js";

// ── Constants ──────────────────────────────────────────────────────

const VALID_CLASSES = new Set<string>(["EQ", "OPT", "FUT", "FOP", "V"]);
const TOKEN_RE = /^[A-Z0-9_-]+$/;
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;
const RIGHT_RE = /^[CP]$/;

// ── Parse ──────────────────────────────────────────────────────────

export function parse(s: string): ParsedSymbol {
  if (typeof s !== "string" || s.length === 0) {
    throw new Error("Symbol must be a non-empty string");
  }

  const parts = s.split(":");
  const cls = parts[0] as SymbolClass;

  if (!VALID_CLASSES.has(cls)) {
    throw new Error(`Unknown symbol class: ${cls}`);
  }

  switch (cls) {
    case "EQ": {
      if (parts.length !== 2)
        throw new Error(`EQ symbol must have 2 parts, got ${parts.length}: ${s}`);
      const ticker = parts[1]!;
      if (!TOKEN_RE.test(ticker)) throw new Error(`Invalid ticker: ${ticker}`);
      return { class: "EQ", ticker };
    }

    case "OPT": {
      if (parts.length !== 5)
        throw new Error(`OPT symbol must have 5 parts, got ${parts.length}: ${s}`);
      const [, root, expiry, right, strikeStr] = parts;
      if (!TOKEN_RE.test(root!)) throw new Error(`Invalid root: ${root}`);
      if (!DATE_RE.test(expiry!)) throw new Error(`Invalid expiry date: ${expiry}`);
      if (!RIGHT_RE.test(right!)) throw new Error(`Invalid right: ${right}`);
      const strike = Number(strikeStr);
      if (!Number.isFinite(strike) || strike <= 0) throw new Error(`Invalid strike: ${strikeStr}`);
      return { class: "OPT", root: root!, expiry: expiry!, right: right! as "C" | "P", strike };
    }

    case "FUT": {
      if (parts.length !== 3)
        throw new Error(`FUT symbol must have 3 parts, got ${parts.length}: ${s}`);
      const [, root, contract] = parts;
      if (!TOKEN_RE.test(root!)) throw new Error(`Invalid root: ${root}`);
      if (
        !TOKEN_RE.test(contract!) &&
        !DATE_RE.test(contract!) &&
        !/^\d{4}-\d{2}$/.test(contract!)
      ) {
        throw new Error(`Invalid contract: ${contract}`);
      }
      return { class: "FUT", root: root!, contract: contract! };
    }

    case "FOP": {
      if (parts.length !== 6)
        throw new Error(`FOP symbol must have 6 parts, got ${parts.length}: ${s}`);
      const [, root, contract, expiry, right, strikeStr] = parts;
      // Allow both YYYY-MM and YYYY-MM-DD for contract
      if (!TOKEN_RE.test(root!)) throw new Error(`Invalid root: ${root}`);
      if (!DATE_RE.test(expiry!)) throw new Error(`Invalid expiry date: ${expiry}`);
      if (!RIGHT_RE.test(right!)) throw new Error(`Invalid right: ${right}`);
      const strike = Number(strikeStr);
      if (!Number.isFinite(strike) || strike <= 0) throw new Error(`Invalid strike: ${strikeStr}`);
      return {
        class: "FOP",
        root: root!,
        contract: contract!,
        expiry: expiry!,
        right: right! as "C" | "P",
        strike,
      };
    }

    case "V": {
      if (parts.length !== 2)
        throw new Error(`V symbol must have 2 parts, got ${parts.length}: ${s}`);
      const label = parts[1]!;
      // V symbols allow lowercase per TDD: [a-z0-9_-]+ for labels like "regime-spx-vol"
      if (!/^[A-Za-z0-9_-]+$/.test(label)) throw new Error(`Invalid virtual label: ${label}`);
      return { class: "V", label };
    }

    default:
      throw new Error(`Unhandled symbol class: ${cls}`);
  }
}

// ── Format ─────────────────────────────────────────────────────────

function formatStrike(strike: number): string {
  // No trailing zeros: 500, 4787.5
  return String(strike);
}

export function format(sym: ParsedSymbol): string {
  switch (sym.class) {
    case "EQ":
      return `EQ:${sym.ticker}`;
    case "OPT":
      return `OPT:${sym.root}:${sym.expiry}:${sym.right}:${formatStrike(sym.strike)}`;
    case "FUT":
      return `FUT:${sym.root}:${sym.contract}`;
    case "FOP":
      return `FOP:${sym.root}:${sym.contract}:${sym.expiry}:${sym.right}:${formatStrike(sym.strike)}`;
    case "V":
      return `V:${sym.label}`;
  }
}

// ── Subject Tokens (for NATS subject construction) ─────────────────

export function toSubjectTokens(sym: ParsedSymbol): string[] {
  // NATS uses "." as subject delimiter, so tokens must not contain dots.
  // Fractional strikes (e.g., 4787.5) use "_" as the decimal separator in subjects.
  function strikeToken(strike: number): string {
    return formatStrike(strike).replace(".", "_");
  }

  switch (sym.class) {
    case "EQ":
      return ["EQ", sym.ticker];
    case "OPT":
      return ["OPT", sym.root, sym.expiry, sym.right, strikeToken(sym.strike)];
    case "FUT":
      return ["FUT", sym.root, sym.contract];
    case "FOP":
      return ["FOP", sym.root, sym.contract, sym.expiry, sym.right, strikeToken(sym.strike)];
    case "V":
      return ["V", sym.label];
  }
}

// ── Filename helpers ───────────────────────────────────────────────

export function toFilename(symbolStr: string): string {
  return symbolStr.replace(/:/g, "-");
}

export function fromFilename(filename: string): string {
  // Reverse of toFilename: first token before first dash is the class
  // EQ-SPY → EQ:SPY, OPT-SPY-2026-01-16-C-500 → OPT:SPY:2026-01-16:C:500
  // This is lossy for dates (dashes are ambiguous). Use the class to guide reconstruction.
  const firstDash = filename.indexOf("-");
  if (firstDash === -1) return filename;

  const cls = filename.slice(0, firstDash);
  const rest = filename.slice(firstDash + 1);

  switch (cls) {
    case "EQ":
      return `EQ:${rest}`;
    case "OPT": {
      // SPY-2026-01-16-C-500 → SPY:2026-01-16:C:500
      const m = rest.match(/^([A-Z0-9_-]+?)-(\d{4}-\d{2}-\d{2})-([CP])-(.+)$/);
      if (m) return `OPT:${m[1]}:${m[2]}:${m[3]}:${m[4]}`;
      break;
    }
    case "FUT": {
      // ES-2026-06 → FUT:ES:2026-06
      const m = rest.match(/^([A-Z0-9_-]+?)-(\d{4}-\d{2})$/);
      if (m) return `FUT:${m[1]}:${m[2]}`;
      break;
    }
    case "FOP": {
      const m = rest.match(/^([A-Z0-9_-]+?)-(\d{4}-\d{2})-(\d{4}-\d{2}-\d{2})-([CP])-(.+)$/);
      if (m) return `FOP:${m[1]}:${m[2]}:${m[3]}:${m[4]}:${m[5]}`;
      break;
    }
    case "V":
      return `V:${rest}`;
  }

  // Fallback: naive replacement (will break on dates but better than nothing)
  return filename.replace(/-/g, ":");
}
